src/rust/ide/src/transport/web.rs

```diff
@@ -98,60 +103,198 @@ impl State {
+    #[derive(Clone,Copy,Debug)]
+    pub enum Message{}
+    impl Event for Message {
+        type Interface = web_sys::MessageEvent;
+        type Target = web_sys::WebSocket;
+        const NAME:&'static str = "message";
+    }
+
+    /// Represents WebSocket.error event.
+    #[derive(Clone,Copy,Debug)]
+    pub enum Error{}
+    impl Event for Error {
+        type Interface = web_sys::Event;
+        type Target = web_sys::WebSocket;
+        const NAME:&'static str = "error";
+    }
 }
 
-impl WebSocket {
+
+
+// =============
+// === Model ===
+// =============
+
+/// An owning wrapper over JS `WebSocket` object and callbacks to its signals.
+#[derive(Derivative)]
+#[derivative(Debug)]
+#[allow(missing_docs)]
+pub struct Model {
+    // === User-provided callbacks ===
+    pub on_close          : Slot<event::Close>,
+    pub on_message        : Slot<event::Message>,
+    pub on_open           : Slot<event::Open>,
+    pub on_error          : Slot<event::Error>,
+
+
+    // === Internal ===
+    pub logger            : Logger,
+    pub ws                : web_sys::WebSocket,
+    /// Special callback on "close" event. Must be invoked after `on_close`.
+    pub on_close_internal : Slot<event::Close>,
+    /// When enabled, the WS will try to automatically reconnect whenever connection is lost.
+    pub reconnect         : bool,
+}
+
+impl Model {
     /// Wraps given WebSocket object.
-    pub fn new
-    (ws:web_sys::WebSocket, parent:impl AnyLogger, name:impl AsRef<str>) -> WebSocket {
+    pub fn new(ws:web_sys::WebSocket, logger:Logger) -> Model {
         ws.set_binary_type(BinaryType::Arraybuffer);
-        WebSocket {
+        Model {
+            on_close          : Slot::new(&ws, &logger),
+            on_message        : Slot::new(&ws, &logger),
+            on_open           : Slot::new(&ws, &logger),
+            on_error          : Slot::new(&ws, &logger),
+            on_close_internal : Slot::new(&ws, &logger),
+            reconnect         : true,
+            logger,
             ws,
-            logger     : Logger::sub(parent,name),
-            on_message : default(),
-            on_close   : default(),
-            on_open    : default(),
-            on_error   : default(),
         }
     }
 
+    /// Close the socket.
+    pub fn close(&mut self, reason:&str) -> Result<(),JsValue> {
+        self.ws.close_with_code_and_reason(1000,reason)?;
+        self.clear_callbacks();
+        Ok(())
+    }
+
+    /// Clear all the available callbacks.
+    pub fn clear_callbacks(&mut self) {
+        let Self{on_close,on_error,on_message,on_open,on_close_internal,..} = self;
+        // We don't care if removing actually removed anything.
+        // If callbacks were not set, then they are clear from the start.
+        on_close.clear_callback();
+        on_error.clear_callback();
+        on_message.clear_callback();
+        on_open.clear_callback();
+        on_close_internal.clear_callback()
```

Comment:
What this syntax is better in instead of just using `self.` as the beginning of the lines? I feel it confusing tbh - reading the later lines I need to remember which things are from `self`. Moreover, it is longer than just writing `self` everywhere , so I don't see a reason for it tbh.

Reply:
The purpose was to have compiler bring our attention here each time when fields are added.
I just forgot to remove `..`.
Fixed and clarified in the comment.